examples/05.HTTP_SERVER/http_server.cc

```diff
@@ -0,0 +1,209 @@
+
+static char reply[] =
+  "HTTP/1.1 200 OK\r\n"
+  "Content-type: text/html\r\n"
+  "Connection: close\r\n"
+  "\r\n"
+  "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">"
+  "<html>"
+  "<head><title>Hello from CHERIoT!</title></head>"
+  "<body><h1>It works!</h1><p>Served from a CHERIoT device.</p></body>"
+  "</html>\n";
+
+/**
+ * Maximum number of clients the server will server before shutting down. This
+ * is useful to check that the server can handle multiple clients before
+ * terminating.
+ */
+static const uint16_t MaxClients = 10;
+
+/**
+ * This example server is written to showcase the network stack server API, but
+ * also the network stack restart. If one of the network stack APIs fails in a
+ * way likely caused by a crash, the server will wait this small delay before
+ * retrying to give time to the network stack to reset.
+ */
+static const uint16_t RestartDelay = 100; // in ticks
+
+void __cheri_compartment("http_server_example") example()
+{
+	network_start();
+
+	auto heapAtStart = heap_quota_remaining(TEST_MALLOC);
+
+	uint16_t clientsCounter = 0;
+
+	Debug::log("Starting the server.");
+
+	while (clientsCounter < MaxClients)
+	{
+		Debug::log("Creating a listening socket.");
+		Timeout unlimited{UnlimitedTimeout};
+		auto    socket = network_socket_listen_tcp(
+		     &unlimited, TEST_MALLOC, STATIC_SEALED_VALUE(HTTPPort));
+
+		if (!Capability{socket}.is_valid())
+		{
+			Debug::log("Failed to create a listening socket.");
+			// This may have failed because of a network stack
+			// crash. Sleep a little bit to enable a reset.
+			Timeout sleep{RestartDelay};
+			thread_sleep(&sleep);
+			continue;
+		}
+
+		Debug::log("Listening on port {}...", LISTEN_PORT);
+		while (clientsCounter < MaxClients)
+		{
+			NetworkAddress clientAddress = {0};
+			uint16_t       clientPort    = 0;
+
+			auto clientSocket = network_socket_accept_tcp(
+			  &unlimited, TEST_MALLOC, socket, &clientAddress, &clientPort);
+
+			if (!Capability{clientSocket}.is_valid())
+			{
+				Debug::log("Failed to establish a connection.");
+				Timeout sleep{RestartDelay};
+				thread_sleep(&sleep);
+				break;
+			}
+
+			if (!UseIPv6)
+			{
+				Debug::log("Established a connection with {}.{}.{}.{}:{}",
+				           static_cast<int>(clientAddress.ipv4) & 0xff,
+				           static_cast<int>(clientAddress.ipv4 >> 8) & 0xff,
+				           static_cast<int>(clientAddress.ipv4 >> 16) & 0xff,
+				           static_cast<int>(clientAddress.ipv4 >> 24) & 0xff,
+				           clientPort);
+			}
+			else
+			{
+				Debug::log("Established a connection.");
+			}
+
+			clientsCounter++;
+
+			auto [received, buffer] =
+			  network_socket_receive(&unlimited, TEST_MALLOC, clientSocket);
+
+			// For this simple server, we do not care about what the client
+			// sent (we will always serve the same content).
+			int ret = heap_free(TEST_MALLOC, buffer);
+			if (ret != 0)
+			{
+				// This may happen if the network stack crashed.
```

Comment:
To check my understanding: because it may have been freed out from under us, and we aren't holding a claim on it?  That is, this isn't a memory leak.

Reply:
Actually this may simply happen because the TCP/IP stack crashed during `network_socket_receive`, resulting in `buffer` being a nullptr, or... any other untagged thing. `heap_free` will then return an error when we pass it this invalid thing.

Thank you for this comment, this is worth documenting!